app/controllers/locations_controller.rb

```diff
@@ -27,14 +27,40 @@ def index
       search_params["main_category"] = ""
     end
 
-    if !params["categories"] and !search_params["main_category"].empty?
-      if params["main_category_id"].present?
-        search_params["categories"] = [params["main_category_id"]]
-      end
+    @matched_category = match_keyword_to_category(search_params[:keyword])
+
+    if @matched_category
+      search_params[:categories] = [@matched_category.id]
+      search_params[:main_category] = @matched_category.parent&.name || @matched_category.name
+      @main_category_selected_name = search_params[:main_category]
+      @main_category_selected_id = @matched_category.parent&.id || @matched_category.id
+      search_params[:keyword] = nil
     else
-      search_params["categories"] = params["categories_ids"]
-    end
+      @main_category_selected_name = ""
+      @main_category_selected_id = nil
+      search_params[:categories] = []
+
+      if search_params[:keyword].present?
+        search_params[:main_category] = ""
```

Comment:
Im not sure that we should be cleaning the category and the main category when there is a search param. In the production version it looks for the keywords within the constrains of the main_category and the category